Framework/CurveFitting/src/Functions/Meier.cpp

```diff
@@ -0,0 +1,184 @@
+
+  const double OmegaD = 2 * M_PI * FreqD;
+  const double OmegaQ = 2 * M_PI * FreqQ;
+
+  const size_t size = int(J2 + 2);
+  sinAlphaSquared.resize(size);
+  cosAlphaSquared.resize(size);
+  sin2AlphaSquared.resize(size);
+  cos2AlphaSquared.resize(size);
+  negativeLambda.resize(size);
+  positiveLambda.resize(size);
+  std::valarray<double> Wm(size);
+
+  for (size_t i = 0; i < size; i++) {
+    const double m = static_cast<double>(i) - J;
+    const double q1 = (OmegaQ + OmegaD) * (2 * m - 1);
+    const double q2 = OmegaD * std::sqrt(J * (J + 1) - m * (m - 1));
+    const double qq = pow(q1, 2) + pow(q2, 2);
+    const double q3 = OmegaQ * (2 * pow(m, 2) - 2 * m + 1) + OmegaD;
+
+    Wm[i] = std::sqrt(qq);
+
+    if (static_cast<double>(i) < (J2 + 1)) {
+      positiveLambda[i] = 0.5 * (q3 + Wm[i]);
+    } else {
+      positiveLambda[i] = OmegaQ * pow(J, 2) - OmegaD * J;
+    }
+
+    if (i > 0) {
+      negativeLambda[i] = 0.5 * (q3 - Wm[i]);
+    } else {
+      negativeLambda[i] = OmegaQ * pow(J, 2) - OmegaD * J;
+    }
+
+    if (qq > 0) {
+      cos2AlphaSquared[i] = pow(q1, 2) / qq;
+    } else {
+      cos2AlphaSquared[i] = 0;
+    }
+    sin2AlphaSquared[i] = 1 - cos2AlphaSquared[i];
+
+    cosAlphaSquared[i] = 0.5 * (1 + std::sqrt(cos2AlphaSquared[i]));
+    sinAlphaSquared[i] = 1 - cosAlphaSquared[i];
+  }
+}
+
+/**
+ * Calculates the polarization in the x direction
+ * @param Px :: the polarization in the x direction
+ * @param xValues :: input x values
+ * @param sinAlphaSquared :: sin of alpha squared
+ * @param cosAlphaSquared :: cos of alpha squared
+ * @param sin2AlphaSquared :: sin of 2*alpha squared
+ * @param cos2AlphaSquared :: cos of 2*alpha squared
+ * @param positiveLambda :: negative lambda
+ * @param negativeLambda :: positive lambda
+ * @param J2 :: 2 * total angular momentum quantum number multiplied by 2
+ */
+void Meier::calculatePx(std::valarray<double> &Px, const std::valarray<double> &xValues,
+                        const std::valarray<double> &sinAlphaSquared, const std::valarray<double> &cosAlphaSquared,
+                        const std::valarray<double> &negativeLambda, const std::valarray<double> &positiveLambda,
+                        const double &J2) const {
+  std::valarray<double> tx(xValues.size());
+  for (int i = 0; i < int(J2) + 1; i++) {
+    const std::valarray<double> a =
+        cosAlphaSquared[i + 1] * sinAlphaSquared[i] * std::cos((positiveLambda[i + 1] - positiveLambda[i]) * xValues);
+    const std::valarray<double> b =
+        cosAlphaSquared[i + 1] * cosAlphaSquared[i] * std::cos((positiveLambda[i + 1] - negativeLambda[i]) * xValues);
+    const std::valarray<double> c =
+        sinAlphaSquared[i + 1] * sinAlphaSquared[i] * std::cos((negativeLambda[i + 1] - positiveLambda[i]) * xValues);
+    const std::valarray<double> d =
+        sinAlphaSquared[i + 1] * cosAlphaSquared[i] * std::cos((negativeLambda[i + 1] - negativeLambda[i]) * xValues);
+    tx = tx + a + b + c + d;
```

Comment:
what are a, b, c, d?

Reply:
could we use ``` tx += ...```

Reply:
They are the four terms in this giant summation
![image](https://user-images.githubusercontent.com/44416400/214889047-bfa1e160-d341-4ff5-b5a0-ece312b0e923.png)